# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/CodeGenerator.cpp
# Commit: 27db0e6a7010
# Full Hash: 27db0e6a7010f290ac14e5d260c0b4fe48442486
# Author: Lars T Hansen <lhansen@mozilla.com>
# Date: 2021-10-18 16:09:29
# Regressor Bug: 1727084
# File Overlap Count: 6
# Description:
#   Bug 1727084 - Memory64 - Ion operations. r=yury
#   
#   Add memory64 generalizations for Ion for all operations (except bulk
#   memory, done in an earlier patch).
#   
# ==============================================================================

diff -r 1ab8383d17ee -r 27db0e6a7010 js/src/jit/CodeGenerator.cpp
--- a/js/src/jit/CodeGenerator.cpp	Mon Oct 18 07:31:56 2021 +0000
+++ b/js/src/jit/CodeGenerator.cpp	Mon Oct 18 07:31:56 2021 +0000
@@ -14851,14 +14851,9 @@
 void CodeGenerator::visitWasmBoundsCheck64(LWasmBoundsCheck64* ins) {
   const MWasmBoundsCheck* mir = ins->mir();
   Label ok;
-#ifdef JS_64BIT
   Register64 ptr = ToRegister64(ins->ptr());
   Register64 boundsCheckLimit = ToRegister64(ins->boundsCheckLimit());
   masm.wasmBoundsCheck64(Assembler::Below, ptr, boundsCheckLimit, &ok);
-#else
-  // 64-bit bounds checks are used only on 64-bit systems.
-  MOZ_CRASH("Should not happen");
-#endif
   masm.wasmTrap(wasm::Trap::OutOfBounds, mir->bytecodeOffset());
   masm.bind(&ok);
 }
@@ -14872,6 +14867,20 @@
   masm.bind(&ok);
 }
 
+void CodeGenerator::visitWasmAlignmentCheck64(LWasmAlignmentCheck64* ins) {
+  const MWasmAlignmentCheck* mir = ins->mir();
+  Register64 ptr = ToRegister64(ins->ptr());
+#ifdef JS_64BIT
+  Register r = ptr.reg;
+#else
+  Register r = ptr.low;
+#endif
+  Label ok;
+  masm.branchTestPtr(Assembler::Zero, r, Imm32(mir->byteSize() - 1), &ok);
+  masm.wasmTrap(wasm::Trap::UnalignedAccess, mir->bytecodeOffset());
+  masm.bind(&ok);
+}
+
 void CodeGenerator::visitWasmLoadTls(LWasmLoadTls* ins) {
   switch (ins->mir()->type()) {
     case MIRType::Pointer: